File: src/main/java/rewriteRules/SplitXOfY.java
package rewriteRules;

import structure.Chunk;
import structure.RewriteCaption;
import structure.RewriteRule;
import utils.ArrayListUtils;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;

/**
 * Split "X of Y"
 * @author aylai2
 */
public class SplitXOfY {

    /**
     * Find "X of Y" cases and drop "X of" and "of Y" where applicable
     * @param cap Caption to be modified
     * @return Modified caption with applicable rules added
     */
    public static RewriteCaption applyRule(RewriteCaption cap) {
        HashSet<Integer> toRemoveRoots = new HashSet<>();
        ArrayList<String> newRoots = new ArrayList<>();
        for (int rootIdx = 0; rootIdx < cap.getRoots().size(); rootIdx++) {
            Chunk sent = cap.getRoots().get(rootIdx);
            String origRootStr = sent.toBareString();
            ArrayList<Chunk> chunks = new ArrayList<>();
            if (sent.getType().equals("SENT")) {
                chunks = sent.getChunks();
            } else {
                chunks.add(sent);
            }
            for (Chunk c : chunks) {
                // look for "[EN [ ] [PP of ] [ ] ]"
                if (!c.getType().equals("EN")) {
                    continue;
                }
                // second chunk in EN must be [PP of ]
                ArrayList<Chunk> children = c.getChunks();
                if (children.size() == 3 && children.get(1).getType().equals("PP") && children.get(1).toBareString().equals("of")) {
                    Chunk chunkX = children.get(0); // first NP chunk
                    Chunk chunkOfPP = children.get(1); // PP "of" chunk
                    Chunk chunkY = children.get(2); // second NP chunk
                    String headStr = c.getChunkHead();
                    // generate "X of Y" -> "X"
                    ArrayList<String> left = new ArrayList<>();
                    ArrayList<String> right = new ArrayList<>();
                    // if "body of water", do not generate rule "body of water" -> "body"
                    // don't generate "sort/kind/type of Y" -> "sort/kind/type"
                    if (!headStr.equals("body/water") && !headStr.startsWith("sort/") && !headStr.startsWith("kind/") && !headStr.startsWith("type/") && !headStr.equals("bunch/person")) {
                        Chunk newSent = new Chunk(sent.toString());
                        // left side of rule is: [EN [X_NP [PP [Y] ] where closing bracket belongs to EN chunk
                        right.add(Integer.toString(c.getFirstToken().getId()));
                        right.add(Integer.toString(chunkX.getFirstToken().getId()));
                        right.add(chunkOfPP.toString());
                        right.add(chunkY.toString());
                        right.add(Integer.toString(c.getLastToken().getId()));
                        // right side of rule is: [EN [X_NP ] where closing bracket belongs to EN chunk
                        left.add(Integer.toString(c.getFirstToken().getId()));
                        left.add(Integer.toString(chunkX.getFirstToken().getId()));
                        left.add(Integer.toString(c.getLastToken().getId()));
                        newSent.removeChunk(chunkY, true);
                        newSent.removeChunk(chunkOfPP, true);
                        // create new rewrite rule
                        if (!newSent.toBareString().equals("")) {
                            cap.addRule(new RewriteRule(cap.getRules().size(), left, right, "+ofY"), origRootStr);
                            // check if we need to create rewrite rules for hyponyms
                            for (Chunk yInner : chunkY.getAllChunks()) {
                                String yInnerStr = yInner.toString();
                                if (cap.hasHyponyms(yInnerStr)) {
                                    // create new rule
                                    String rightStr = ArrayListUtils.stringListToString(right, " ");
                                    HashSet<String> hyponyms = cap.getHyponyms(yInnerStr);
                                    for (String hyponym : hyponyms) {
                                        String rightStrNew = rightStr.replace(yInnerStr, hyponym);
                                        cap.addRule(new RewriteRule(cap.getRules().size(), ArrayListUtils.stringListToString(left, " "), rightStrNew, "+ofY"), origRootStr);
                                    }
                                }
                            }
                            newRoots.add(newSent.toString());
                            toRemoveRoots.add(rootIdx);
                        }
                    }

                    // generate "X of Y" -> "Y"
                    Chunk newSent = new Chunk(sent.toString());
                    left = new ArrayList<>();
                    right = new ArrayList<>();
                    // left side of rule is: [EN [X] [PP [Y_NP ] where closing bracket belongs to EN chunk
                    right.add(Integer.toString(c.getFirstToken().getId()));
                    right.add(chunkX.toString());
                    right.add(chunkOfPP.toString());
                    right.add(Integer.toString(chunkY.getFirstToken().getId()));
                    right.add(Integer.toString(c.getLastToken().getId()));
                    // right side of rule is: [EN [Y_NP ] where closing bracket belongs to EN chunk
                    left.add(Integer.toString(c.getFirstToken().getId()));
                    left.add(Integer.toString(chunkY.getFirstToken().getId()));
                    left.add(Integer.toString(c.getLastToken().getId()));
                    newSent.removeChunk(chunkOfPP, true);
                    newSent.removeChunk(chunkX, true);
                    // create new rewrite rule
                    if (!newSent.toString().equals("")) {
                        cap.addRule(new RewriteRule(cap.getRules().size(), left, right, "+Xof"), origRootStr);
                        // check if we need to create rewrite rules for hyponyms
                        for (Chunk xInner : chunkX.getAllChunks()) {
                            String xInnerStr = xInner.toString();
                            if (cap.hasHyponyms(xInnerStr)) {
                                // create new rule
                                String rightStr = ArrayListUtils.stringListToString(right, " ");
                                HashSet<String> hyponyms = cap.getHyponyms(xInnerStr);
                                for (String hyponym : hyponyms) {
                                    String rightStrNew = rightStr.replace(xInnerStr, hyponym);
                                    cap.addRule(new RewriteRule(cap.getRules().size(), ArrayListUtils.stringListToString(left, " "), rightStrNew, "+Xof"), origRootStr);
                                }
                            }
                        }
                        newRoots.add(newSent.toString());
                        toRemoveRoots.add(rootIdx);
                    }
                }
            }
        }
        ArrayList<Integer> removeRootsList = new ArrayList(toRemoveRoots);
        Collections.sort(removeRootsList);
        for (int listIdx = removeRootsList.size() - 1; listIdx >= 0; listIdx--) {
            int idxRemove = removeRootsList.get(listIdx);
            cap.removeRoot(idxRemove);
        }
        for (String root : newRoots) {
            cap.addRoot(root);
        }
        return cap;

    }

}

File: preprocessing/graph/scripts/splitXofY.pl
#!/usr/bin/perl

use FindBin;
use lib "$FindBin::Bin";
use simple;

@dep = ();
@X = ();
@Y = ();
@type = ();
$n = 0;
open(file, $ARGV[0]);
while (<file>) {
	chomp($_);
	@ax = split(/\t/, $_);
	if ($#ax == 1) {
		$ax[2] = "";
	}
	# this is a caption
	if ($#ax == 2) {
		@ay = split(/ /, $ax[2]);
		($next, $prev) = breakSlash(\@ay, 1);

		# look for an EN chunk with a "[PP of ]" as its second chunk
		for ($i = 0; $i <= $#ay; $i += $next->[$i]) {
			if ($ay[$i]->[1] eq "[EN") {
				$j = $i + 1;
				if ($ay[$j + $next->[$j] + 0]->[1] eq "[PP" &&
					$ay[$j + $next->[$j] + 1]->[1] eq "of" &&
					$ay[$j + $next->[$j] + 2]->[1] eq "]") {
					# $j1 - first NP chunk (X)
					# $j2 - [PP of ] chunk
					# $j3 - second NP chunk (Y)
					# $j4 - end of the EN chunk
					$j1 = $j;
					$j2 = $j1 + $next->[$j1];
					$j3 = $j2 + $next->[$j2];
					$j4 = $j3 + $next->[$j3];

					# check if it's "body of water".  We don't want the
					# "body of water" -> "body rule"
					$x = entityString(\@ay, $next, $i);
					if ($x ne "body/water") {
						# generated the X of Y -> X rule
						# @aX - left side of rule
						# @aY - right side of rule
						@aX = ();
						@aY = ();
						push(@aX, $ay[$i]->[0]);
						push(@aY, $ay[$i]->[0]);
						push(@aX, $ay[$j1]->[0]);
						push(@aY, $ay[$j1]->[0]);
						push(@aX, $ay[$j2]->[0]);
						for ($k = $j3; $k < $j4; $k++) {
							push(@aX, $ay[$k]->[0]);
						}
						push(@aX, $ay[$j4]->[0]);
						push(@aY, $ay[$j4]->[0]);
						addTransformation("", join(" ", @aX), join(" ", @aY), "-ofY", \@dep, \@X, \@Y, \@type, \$n);
					}

					# generated the X of Y -> Y rule
					# @aX - left side of rule
					# @aY - right side of rule
					@aX = ();
					@aY = ();
					push(@aX, $ay[$i]->[0]);
					push(@aY, $ay[$i]->[0]);
					for ($k = $j1; $k < $j2; $k++) {
						push(@aX, $ay[$k]->[0]);
					}
					push(@aX, $ay[$j2]->[0]);
					push(@aX, $ay[$j3]->[0]);
					push(@aY, $ay[$j3]->[0]);
					push(@aX, $ay[$j4]->[0]);
					push(@aY, $ay[$j4]->[0]);
					addTransformation("", join(" ", @aX), join(" ", @aY), "-Xof", \@dep, \@X, \@Y, \@type, \$n);
				}
			}
		}

		printSentence($ax[0], $ax[1], \@ay, \@dep, \@X, \@Y, \@type, $n);

		@dep = ();
		@X = ();
		@Y = ();
		@type = ();
		$n = 0;
	# this is a rule - add it if the index is correct ($n)
	} elsif ($#ax == 4) {
		if ($ax[0] == $n) {
			$dep[$ax[0]] = $ax[1];
			$X[$ax[0]] = $ax[2];
			$Y[$ax[0]] = $ax[3];
			$type[$ax[0]] = $ax[4];
			$n++;
		}
	}
}
close(file);
